### ISSUE TITLE
Method returning self type in generic class erroneously fails type-check due to "unknown" return type

### ISSUE THREAD

**Issue by WasabiFan** [microsoft/pyright#3004]
**Describe the bug**

I have a method in a generic class, `Foo`. This method returns `Foo[T]` for arbitrary `T`. It does so by returning the result of the expression `Foo[T]()`. This fails type-check in strict mode because, according to Pyright, `Foo[T]()` evaluates to an unknown type.

**To Reproduce**

Type-check the included code snippet with Pyright version 1.1.219 (or 1.1.218. which was the most current version when I started prepping this repro case half an hour ago 😆 ). **Use `typeCheckingMode = "strict"` (or enable the "no unknown types" flag).**

You will see that it fails the type-check. Both the function which directly returns the expression, and the one which assigns it to a variable, fail.

To make matters more curious, uncomment the empty constructor. Re-run type-check. The function which returns through an intermediate variable now _passes_ typecheck, while the other one still fails.

In case this wasn't sufficiently entertaining, downgrade to Pyright version 1.1.180. With the constructor commented out, the behavior is the same as before. With the constructor included, **the program passes type-check**. In other words, I view this as a regression -- albeit, there seems to still be a bug present in the older version.

**Expected behavior**

The functions fully specify that the returned value is of the same generic parameter type as the receiver generic class. Thus, I would expect the type-check to pass.

**Screenshots or Code**

```python
from typing import Generic, TypeVar

T = TypeVar("T")

class Foo(Generic[T]):
    # Try uncommenting this constructor -- return_from_variable now passes type-check!
    # def __init__(self):
    #     pass

    def return_directly(self) -> "Foo[T]":
        return Foo[T]()

    def return_from_variable(self) -> "Foo[T]":
        value = Foo[T]()
        reveal_type(value)
        return value

x = Foo[int]()
returned_directly = x.return_directly()
returned_from_variable = x.return_from_variable()

reveal_type(returned_directly)
reveal_type(returned_from_variable)
```

With Pyright 1.1.219, I get:

```
path/to/test.py
  path/to/test.py:11:16 - error: Return type, "Foo[Unknown]", is partially unknown (reportUnknownVariableType)
  path/to/test.py:14:9 - error: Type of "value" is partially unknown
    Type of "value" is "Foo[Unknown]" (reportUnknownVariableType)
  path/to/test.py:15:21 - information: Type of "value" is "Foo[Unknown]"
  path/to/test.py:16:16 - error: Return type, "Foo[Unknown]", is partially unknown (reportUnknownVariableType)       
  path/to/test.py:22:13 - information: Type of "returned_directly" is "Foo[int]"
  path/to/test.py:23:13 - information: Type of "returned_from_variable" is "Foo[int]"
3 errors, 0 warnings, 3 informations
```

**VS Code extension or command-line**

Pyright `1.1.219` and `1.1.180`, as described above. `typeCheckingMode` is `strict`.

**Additional context**

None.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.220, which I just published. It will also be included in the next release of pylance.

**Comment by erictraut** [comment]
Sorry, I missed that case. Looks like there's a second issue related to bidirectional type inference in that case. Reopening the issue.

**Comment by WasabiFan** [comment]
Thank you!

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.221, which I just published. It will also be included in the next release of pylance.

**Comment by WasabiFan** [comment]
Thanks! I can confirm this seems to fix the problem on my codebase.
